pkg/k8s/watchers/cilium_endpoint_batch.go

```diff
@@ -0,0 +1,202 @@
+	"github.com/cilium/cilium/pkg/k8s/informer"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/sirupsen/logrus"
+)
+
+func (k *K8sWatcher) ciliumEndpointBatchInit(client *k8s.K8sCiliumClient, asyncControllers *sync.WaitGroup) {
+
+	log.Info("Initializing CEB controller")
+	var once sync.Once
+	_, cebController := informer.NewInformer(
+		cache.NewListWatchFromClient(client.CiliumV2alpha1().RESTClient(),
+			cilium_v2a1.CEBPluralName, v1.NamespaceAll, fields.Everything()),
+		&cilium_v2a1.CiliumEndpointBatch{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if ceb := k8s.ObjToCiliumEndpointBatch(obj); ceb != nil {
+					k.endpointBatchAdded(ceb)
+				}
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				if oldCeb := k8s.ObjToCiliumEndpointBatch(oldObj); oldCeb != nil {
+					if newCeb := k8s.ObjToCiliumEndpointBatch(newObj); newCeb != nil {
+						if oldCeb.DeepEqual(newCeb) {
+							return
+						}
+						k.endpointBatchUpdated(oldCeb, newCeb)
+					}
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if ceb := k8s.ObjToCiliumEndpointBatch(obj); ceb != nil {
+					k.endpointBatchDeleted(ceb)
+				}
+			},
+		},
+		nil,
+	)
+	once.Do(func() {
+		// Signalize that we have put node controller in the wait group
+		// to sync resources.
+		asyncControllers.Done()
+	})
+	k.blockWaitGroupToSyncResources(
+		wait.NeverStop,
+		nil,
+		cebController.HasSynced,
+		k8sAPIGroupCiliumEndpointBatchV2Alpha1,
+	)
+
+	go cebController.Run(wait.NeverStop)
+}
+
+// endpointBatchAdded invoked for newly created CEBs, iterates over CoreCiliumEndpoints
+// packed in ceb, converts coreCEP into types.CEP and call endpointUpdated only for remote node CEPs.
+func (k *K8sWatcher) endpointBatchAdded(ceb *cilium_v2a1.CiliumEndpointBatch) {
+
+	nodeIP := node.GetIPv4().String()
+	if option.Config.EnableIPv6 {
+		nodeIP = node.GetIPv6().String()
+	}
+
+	for i, ep := range ceb.Endpoints {
+		log.WithFields(logrus.Fields{
+			"CebName":          ceb.GetName(),
+			"CoreEndpointName": ep.Name,
+		}).Debug("New CEB added, Calling CoreEndpointUpdate")
+		c := k8s.ConvertCoreCiliumEndpointToTypesCiliumEndpoint(&ceb.Endpoints[i])
+		// Skip processing endpointupdate for local node CEPs
+		if c.Networking != nil && c.Networking.NodeIP == nodeIP {
+			continue
+		}
+		k.endpointUpdated(nil, c)
+	}
+}
+
+// endpointBatchUpdated invoked for modified CEBs, it compares old CEB and new CEB objects
+// determines below things
+// 1) any new core CEPs are packed in CEB
+// 2) any core CEPs are removed from CEB
+// 3) any existing core CEPs are modified in CEB
+// call endpointUpdated/endpointDeleted only for remote node CEPs.
+func (k *K8sWatcher) endpointBatchUpdated(oldCeb, newCeb *cilium_v2a1.CiliumEndpointBatch) {
+
+	nodeIP := node.GetIPv4().String()
+	if option.Config.EnableIPv6 {
+		nodeIP = node.GetIPv6().String()
+	}
+
+	oldCeps := make(map[string]*cilium_v2a1.CoreCiliumEndpoint)
```

Comment:
nit, the `oldCeps` and `newCeps` can be initialized with length of `len(oldCeb.Endpoints)` and `len(newCeb.Endpoints)` respectively.

Reply:
SG. Done.